src/util/hardware.h

```diff
@@ -0,0 +1,74 @@
+// SPDX-FileCopyrightText: none
+// SPDX-License-Identifier: CC0-1.0
+#pragma once
+
+#include <xboxkrnl/xboxkrnl.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "util/output.h"
+
+static inline uint32_t inl(uint16_t port)
+{
+    uint32_t ret;
+    __asm__ __volatile__("inl %w1, %0"
+                         : "=a"(ret)
+                         : "Nd"(port));
+    return ret;
+}
+
+static void getPICVersion(char pic_version[4])
+{
+    ULONG version;
+    HalWriteSMBusValue(0x20, 0x01, FALSE, 0); // reset index to 0 for reading
+    for (unsigned i = 0; i < 3; i++) {
+        HalReadSMBusValue(0x20, 0x01, 0, &version);
+        pic_version[i] = (char)version;
+    }
+    pic_version[3] = '\0';
+}
+
+static const char* getConsoleType(char pic_version[4])
+{
+    // See https://xboxdevwiki.net/Xboxen_Info
+    if (pic_version[0] == 'D' || pic_version[0] == 'B') {
```

Comment:
The linked [website](https://xboxdevwiki.net/Xboxen_Info) also has information for many prototypes and pre-launch date boxes, which I think this code should also detect too. Currently, if `pic_version[0]` is 'D' or 'B' but it's not a devkit or a chihiro system, this code will return "UNKNOWN". There's also the issue that there is a known prototype where `pic_version[0]` is 'A', and the current code will not catch that. While I understand that the note for that entry says the result is from a "random poster", I still feel you could return a model string such as "EVT / DVT1 (uncertain)"

Reply:
I'm assuming `A__` prefix is alpha protocol type machine which is only used in-house experiment or early access to certain partners for maybe feedback and such. Though, we (including nxdk repo), don't support prototypes anyway. Yet... we could simply add `A__` prefix and have it return as "PROTOTYPE` string in case of build xbe for prototype will be supported in the future. (In summary, I'll add prototype check for 'A' prefix soon.)

Return model string isn't intended planned to do as I don't know where to detect it as for specific EVT/DVT#. Plus model string isn't console type which will not return it. And will require a new function to return model string according to hardware specs.

If you looked at X(MODE)2, they are actually debug / dev kits (including chihiro) hardware. Which is what `B__` and `D__` prefixes are. If the kernel flags doesn't set devkit nor chihiro, then technically something went wrong with the kernel or homebrew kernel/patcher. Which is why I let it pass through to return "UNKNOWN".